traits/src/storage.rs

```diff
@@ -0,0 +1,110 @@
+use serde::{de::DeserializeOwned, Serialize};
+
+pub trait Types<const VERSION: usize>: Default {
+    type QueuedProposal: QueuedProposalEntity<VERSION>;
+    type GroupId: GroupIdKey<VERSION>;
+    type ProposalRef: ProposalRefKey<VERSION> + ProposalRefEntity<VERSION>;
+    type TreeSync: TreeSyncEntity<VERSION>;
+    type GroupContext: GroupContextEntity<VERSION>;
+    type InterimTranscriptHash: InterimTranscriptHashEntity<VERSION>;
+    type ConfirmationTag: ConfirmationTagEntity<VERSION>;
+}
+
+pub trait GetError: core::fmt::Debug + std::error::Error + PartialEq {
+    fn error_kind(&self) -> GetErrorKind;
+}
+
+pub trait UpdateError: core::fmt::Debug + std::error::Error + PartialEq {
+    fn error_kind(&self) -> UpdateErrorKind;
+}
+
+pub trait StorageProvider<const VERSION: usize> {
+    // source for errors
+    type GetError: GetError;
+    type UpdateError: UpdateError;
+    type Types: Types<VERSION>;
+
+    // update functions, single and batched
+    fn apply_update(&self, update: Update<VERSION, Self::Types>) -> Result<(), Self::UpdateError>;
+    fn apply_updates(
+        &self,
+        update: Vec<Update<VERSION, Self::Types>>,
+    ) -> Result<(), Self::UpdateError>;
+
+    // getter
+    fn get_queued_proposal_refs(
+        &self,
+        group_id: &<Self::Types as Types<VERSION>>::GroupId,
+    ) -> Result<Vec<<Self::Types as Types<VERSION>>::ProposalRef>, Self::GetError>;
+
+    fn get_queued_proposals(
+        &self,
+        group_id: &<Self::Types as Types<VERSION>>::GroupId,
+    ) -> Result<Vec<<Self::Types as Types<VERSION>>::QueuedProposal>, Self::GetError>;
+
+    fn get_treesync(
```

Comment:
Treesync is not a great name for the outside. We may want to call this just tree.